rake_helper.rb

```diff
@@ -0,0 +1,130 @@
+require 'aws-sdk-lambda'
+require 'aws-sdk-cloudwatchevents'
+require 'yaml'
+
+class RakeHelper
+
+  attr_reader(
+    :travis_branch,
+    :aws_access_key_id,
+    :aws_secret_access_key,
+    :aws_configuration,
+    :lambda_client,
+    :yaml,
+    :lambda_config,
+    :function_name,
+    :event
+  )
+
+  def initialize
+    @travis_branch = ENV["TRAVIS_BRANCH"].upcase
+    @aws_access_key_id = ENV["AWS_ACCESS_KEY_ID_#{travis_branch}"]
+    @aws_secret_access_key = ENV["AWS_SECRET_ACCESS_KEY_#{travis_branch}"]
+    @aws_configuration = aws_configuration = {
+      region: 'us-east-1',
+      access_key_id: aws_access_key_id,
+      secret_access_key: aws_secret_access_key
+    }
+    p 'using configuration: ', aws_configuration
+    if configured?
+      @lambda_client = Aws::Lambda::Client.new(aws_configuration)
+      @lambda_config = yaml["deploy"].find {|conf| conf["function_name"].include? travis_branch.downcase}
+    end
+  end
+
+  def configured?
+    access_key_id && secret_access_key
+  end
+
+
+  def update_lambda_configuration
+    unless configured? && lambda_config
+      p 'insufficient configuration'
+      return nil
+    end
+
+
+    vpc_config = lambda_config["vpc"]
+    environment_variables = lambda_config["environment_variables"]&.map {|str| str.split "="}.to_h
+    layers = lambda_config["layers"]
+    updated_lambda_configuration = [
+      :vpc_config,
+      :environment_variables,
+      :layers
+    ]
+      .select {|key| lambda_config[key]}
+      .map {|key| [key, lambda_config[key]]}
+      .to_h
+    p 'updating_function_configuration with: ', updated_lambda_configuration
+    update_configuration_resp = lambda_client.update_function_configuration(updated_lambda_configuration)
+    p 'update_configuration_resp: ', update_configuration_resp
+  end
+
+  def update_event
+    unless lambda_config["events"]
+      p 'no event config'
+      return nil
+    end
+
+    @event = lambda_config["event"]
+    if event.include?("arn")
+      add_event_source
+    elsif event.include?("SCHEDULE_EXPRESSION")
+      add_cron
+    end
+  end
+
+  def add_event_source
```

Comment:
I don't think I'm following the flow here. If there is an existing event, this looks like it just deletes the existing event and doesn't create a new one? I think if you moved the deletion loop above the creation step you could simplify this so that it just always deletes all existing events and then creates new ones as specified in travis

You could also change this to do event "upserts" but that gets tricky when an event source is deleted but not replaced.

Reply:
It should only be deleting events with arn not equal to the configured event, so e.g. if you change from one kinesis stream to another it will both connect to the new kinesis stream and disconnect from the old one.

Reply:
ah okay, I see that now. But this still  won't update the event settings if the `arn` doesn't change, right? If I change the `batch_size` but nothing else for example, this wouldn't pick up those updates I think

Reply:
Ah, good point. Let me think about that.

Reply:
One option is to always delete all the existing events and create the new one. This has the advantage that it is simple, but it might cause a delay? I am not sure. I've pushed the code for this approach.

Reply:
yeah, that's what I was thinking with my first comment. I don't think it will slow things down, especially since this gets deployed at most a few times a day.

The fix looks good